Feature: Make town bridge max length a function of its population

perezdidac opened the pull request:
Make new town bridge max length a function of its population.

## Motivation / Problem

Really small towns can build disproportionately long bridges, all the way up to 11 tiles. To make it more realistic, the max number of tiles for a built bridge should be a function of its population.

## Description

In order to prevent small towns from building disproportionately long bridges, this change makes the max allowed bridge length a function of its population, while letting every town build at least a town as long as 4 tiles. The max length is defined as a min of 4 plus 1 for each 1000 population, rounded down.

- town of 200 population -> max bridge length of 4
- town of 2000 population -> max bridge length of 6
- town of 10000 population -> max bridge length of 14
- town of 20000 population -> max bridge length of 24

## Limitations

None.

## Checklist for review

This PR is safe, and backwards compatible, since saved games won't be affected other than towns will build bridges with a variable max length as opposed as the current fixed hard coded max length.

perezdidac committed:
Feature: Make maximum length of town bridges depend on population.

James103 commented:
Suggested commit message: "Feature: Make town bridge max length a function of its population"

Also, would it be a good idea to hard cap the town max bridge length at the max bridge length defined in the Advanced Settings?

perezdidac commented:
James103: Yes that is a good idea. We could hard code a limit of 20 probably. I tried different town sizes and bridge lengths and 20 seems too much regardless of the city size. Is there a max bridge length in settings? Any code pointer?

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -1159,19 +1159,24 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
 		/* Bridges starting on flat tiles are only allowed when crossing rivers, rails or one-way roads. */
 		do {
-			if (bridge_length++ >= 4) {
+			if (bridge_length++ >= min_max_bridge_length) {
```
What about `base_bridge_length` instead of `min_max_bridge_length`?

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -1159,19 +1159,24 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
 
+	/* To prevent really small towns from building disproportionately
+	 * long bridges, make the max a function of its population. */
+	int min_max_bridge_length = 4;
```
This only allows a bridge with a span of 3. I suggest increasing this to 5 to allow a bridge over four tracks, which is a much more common track design.

2TallTyler reviewed (commented):
I think 20 tiles is a bit excessive and would lead to large cities building ridiculous bridges. What about keeping the current 11 tile length?

The max bridge length mentioned by James103 is for player-built bridges, which defaults to 64 tiles. Far too long for a town-built bridge, IMHO.

michicc commented:
I like this idea. One idea to consider is to make the length not depend on the population but on the house count as some kind of proxy for "town area".

michicc commented on src/town_cmd.cpp:
```diff
@@ -1159,19 +1159,24 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
+	/* To prevent really small towns from building disproportionately
+	 * long bridges, make the max a function of its population. */
+	int min_max_bridge_length = 4;
+	int max_bridge_length = 0.001 * t->cache.population + min_max_bridge_length;
```
Floating point is a no-no in game code as the rounding might break multi-player. Thankfully, changing it to `/ 1000` is not hard at all.

James103 commented:
But what if the "max bridge length" setting was set to some really low value like 7 tiles (down from the default of 64)?

2TallTyler commented:
@michicc: I like population over number of houses because a small island of skyscrapers may have more population (and thus need for a bridge) than a sprawling town of cottages. This is probably more meaningful with a house NewGRF like Improved Town Layouts where population doesn't necessarily grow linearly with number of houses.

@James103: The max bridge length chosen in the settings is checked by the DoCommand which actually builds the bridge.

TrueBrain commented:
As a small update, we would like to include this PR for our 1.11 release, but there is some small work from our side to do first before we can include it. Mostly, we want to align it with #8473 , and make sure we like the values etc. No work required from your side just yet, but just so you know this is in the works :)

TrueBrain committed:
Change: towns can now bridge 4 rails (up from 3)

Having 4 rails is a pretty common design, and towns now couldn't
bridge out of this common design.

TrueBrain commented:
@perezdidac : I made some minor changes to your code: renamed `min_max` to `base`, remove the floating-point operation, and some minor coding style. Nothing fancy.

I kept it depending on population rather than houses; I think this is a good first iteration to make towns a bit more playful.
